File: apps/mobile/lib/features/slatoki/domain/services/qibla_direction_calculator.dart
import "dart:math" as math;

import "../../../map_discovery/domain/entities/coordinates.dart";

/// Domain Model §4 — the one domain service the Slatoki bounded context
/// owns (Slatoki has no aggregate of its own; see
/// docs/architecture/domain-model.md#4-bounded-context-slatoki). Reuses
/// [Coordinates] from `map_discovery/domain` rather than duplicating a
/// lat/lng value type — that class's own doc comment already frames it as
/// this app's equivalent of the backend's `GeoPosition` shared-kernel
/// value object, just not yet physically relocated to a dedicated shared
/// module (no second feature needed one until this story).
///
/// Pure, stateless, framework-agnostic: no dependency on `flutter_compass`
/// or any device sensor — the live device heading is combined with this
/// calculator's output in the Presentation layer (ADR-0025).
abstract final class QiblaDirectionCalculator {
  /// The Kaaba, Masjid al-Haram, Mecca — standard published WGS84
  /// coordinates (21.422487°N, 39.826206°E), not a project-specific
  /// assumption.
  static const Coordinates kaaba = Coordinates(
    latitude: 21.422487,
    longitude: 39.826206,
  );

  /// The great-circle initial bearing (forward azimuth), in degrees
  /// [0, 360), from [from] to the Kaaba. `0` = due north, `90` = due east.
  static double bearingToMecca(Coordinates from) {
    final double lat1 = from.latitude * math.pi / 180;
    final double lat2 = kaaba.latitude * math.pi / 180;
    final double deltaLng = (kaaba.longitude - from.longitude) * math.pi / 180;

    final double y = math.sin(deltaLng) * math.cos(lat2);
    final double x =
        math.cos(lat1) * math.sin(lat2) -
        math.sin(lat1) * math.cos(lat2) * math.cos(deltaLng);

    final double bearingRadians = math.atan2(y, x);
    final double bearingDegrees = bearingRadians * 180 / math.pi;
    return (bearingDegrees + 360) % 360;
  }
}

File: apps/backend/src/modules/slatoki/domain/services/qibla-direction-calculator.ts
import { GeoPosition } from '../../../../shared-kernel';

/**
 * Domain Model §4 — "stateless: computes bearing to Mecca from a GeoPosition".
 * Documented as a Slatoki-owned domain service for architectural completeness
 * with domain-model.md, but deliberately NOT wired to any controller/endpoint:
 * openapi.yaml's /slatoki/places description is explicit that "Qibla bearing
 * is computed client-side (device compass + great-circle formula) and is
 * intentionally NOT a server endpoint". Kept here, tested, unused by any
 * controller — matching the documented architecture exactly rather than
 * silently dropping a documented domain responsibility or inventing an
 * endpoint the contract doesn't define.
 */
export class QiblaDirectionCalculator {
  /** Kaaba, Mecca (WGS84) — RAH-DOC-005 §2.3. */
  private static readonly MECCA = GeoPosition.of(21.4225, 39.8262);

  /** Initial great-circle bearing in degrees [0, 360), measured clockwise from true north. */
  static bearingFrom(position: GeoPosition): number {
    const lat1 = toRadians(position.lat);
    const lat2 = toRadians(QiblaDirectionCalculator.MECCA.lat);
    const deltaLng = toRadians(QiblaDirectionCalculator.MECCA.lng - position.lng);

    const y = Math.sin(deltaLng) * Math.cos(lat2);
    const x = Math.cos(lat1) * Math.sin(lat2) - Math.sin(lat1) * Math.cos(lat2) * Math.cos(deltaLng);
    const bearing = toDegrees(Math.atan2(y, x));

    return (bearing + 360) % 360;
  }
}

function toRadians(degrees: number): number {
  return (degrees * Math.PI) / 180;
}

function toDegrees(radians: number): number {
  return (radians * 180) / Math.PI;
}
